fetch/cosmwasm_contract/src/access_control.rs

```diff
@@ -0,0 +1,130 @@
+    HumanAddr,
+    StdError,
+    StdResult,
+    Storage,
+    ReadonlyStorage,
+};
+use cosmwasm_storage::{ReadonlyPrefixedStorage, PrefixedStorage};
+
+
+pub static ACCESS_CONTROL_KEY: &[u8] = b"access_control";
+
+pub const OWNER_KEY: &[u8] = b"OWNER_KEY";
+pub const ADMIN_ROLE: &str = "ADMIN_ROLE";
+pub const DELEGATE_ROLE: &str = "DELEGATE_ROLE";
+pub const RELAYER_ROLE: &str = "RELAYER_ROLE";
+
+#[derive(Debug, PartialEq)]
+pub enum AccessRole {
+    Owner,
+    Admin,
+    Delegate,
+    Relayer,
+}
+
+impl AccessRole {
+    fn value(&self) -> &str {
+        match *self {
+            AccessRole::Admin => ADMIN_ROLE,
+            AccessRole::Delegate => DELEGATE_ROLE,
+            AccessRole::Relayer => RELAYER_ROLE,
+            AccessRole::Owner => "",
+        }
+    }
+    
+    fn as_bytes(&self) -> &[u8] {
+        return self.value().as_bytes()
+    }
+    
+    // FIXME(LR) what happen when FromStr treat is in scope?
+    pub fn from_str(s: &str) -> Result<Self, StdError> {
+        match s {
+            ADMIN_ROLE => Ok(AccessRole::Admin),
+            DELEGATE_ROLE => Ok(AccessRole::Delegate),
+            RELAYER_ROLE => Ok(AccessRole::Relayer),
+            _ => Err(StdError::generic_err("Unknow role")),
+        }
+    }
+}
+
+// FIXME(LR) cannot use fn if Trait not in scope, rather annoying
+impl FromStr for AccessRole {
+    type Err = StdError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            ADMIN_ROLE => Ok(AccessRole::Admin),
+            DELEGATE_ROLE => Ok(AccessRole::Delegate),
+            RELAYER_ROLE => Ok(AccessRole::Relayer),
+            _ => Err(StdError::generic_err("Unknow role")),
+        }
+    }
+}
+
+pub fn access_control<S: Storage>(storage: &mut S) -> PrefixedStorage<S> {
+    return PrefixedStorage::new(ACCESS_CONTROL_KEY, storage);
+}
+
+pub fn access_control_read<S: Storage>(storage: &S) -> ReadonlyPrefixedStorage<S> {
+    return ReadonlyPrefixedStorage::new(ACCESS_CONTROL_KEY, storage)
+}
+
+pub fn ac_have_role<S: Storage>(storage: &S, addr: &HumanAddr, role: &AccessRole) -> StdResult<bool> {
+    let ac_store = access_control_read(storage);
+    let addr_roles = ReadonlyPrefixedStorage::new(addr.as_str().as_bytes(), &ac_store);
+    match addr_roles.get(role.as_bytes()) {
+        Some(_) => Ok(true),
+        None => Err(StdError::unauthorized()),
+    }
+}
+
+pub fn ac_add_role<S: Storage>(storage: &mut S, addr: &HumanAddr, role: &AccessRole) -> StdResult<bool> {
+    let already_have_role = ac_have_role(storage, addr, role).unwrap_or(false);
+    if already_have_role {
+        return Err(StdError::generic_err("Already have role")); 
+    }
+    let mut ac_store = access_control(storage);
+    let mut addr_roles = PrefixedStorage::new(addr.as_str().as_bytes(), &mut ac_store);
+    addr_roles.set(role.as_bytes(), b"");
+
+    Ok(true)
+}
+
+pub fn ac_revoke_role<S: Storage>(storage: &mut S, addr: &HumanAddr, role: &AccessRole) -> StdResult<bool> {
```

Comment:
Analogically to my [above comment](https://github.com/fetchai/fetch-ethereum-bridge-v1/pull/20/files#r581856285) - this must be made callable **exclusivelly** by either admin, or by address registered in the role already (= renouncing).

Reply:
same, this is just helper function not a contract call